sw/device/lib/runtime/ibex.h

```diff
@@ -27,15 +27,28 @@ extern const size_t kIbexClockFreqHz;
  *
  * @param cycles the number of cycles to burn.
  */
-inline void ibex_busy_loop(size_t cycles) {
-  size_t out;  // Used to create an inout parameter below.
+inline void ibex_busy_loop(uint64_t cycles) {
+  // Round down |cycles| to closest multiple of 8
+  cycles = cycles & ~7ULL;
+  // Unpack uint64_t into two uint32_ts.
+  uint32_t cycles_lower = cycles;
+  uint32_t cycles_upper = cycles >> 32;
+  // To tell the compiler that |cycles_lower| and |cycles_upper| are modified.
+  uint32_t out_lower, out_upper;
+  // Takes approximately 8 cycles per interation
   asm volatile(
-      "busy_loop%=:"
-      "  nop; nop; nop; nop;"    // 4 cycles.
-      "  addi %1, %1, -8;"       // 1 cycle.
-      "  blez %1, busy_loop%=;"  // 3 cycles.
-      : "=&r"(out)
-      : "0"(cycles));
+      "lower%=:"             // For decrementing |cycles_lower|
+      "  beqz %2, upper%=;"  //   Branch to |upper| if |cycles_lower| is zero.
+      "  addi %2, %2, -8;"   //   Decrement |cycles_lower| by eight.
+      "  j lower%=;"         //   Jump to |lower|.
+      "upper%=:"             // For decrementing |cycles_upper|
+      "  beqz %3, done%=;"   //   Branch to |done| if |cycles_upper| is zero.
+      "  addi %2, %2, -8;"   //   Decrement |cycles_lower| by eight.
+      "  addi %3, %3, -1;"   //   Decrement |cycles_upper| by one, i.e. borrow.
+      "  j lower%=;"         //   Jump to |lower|.
```

Comment:
Does this really add up to 8 cycles? Please have a look at the waveform to confirm.

Reply:
I would note the stall cycles on the branch and jump does depend upon the interconnect. If I remember correctly this is actually different between the FPGA and Verilator versions of earlgrey.

It may be better to simply spin reading mcycle until it hits the appropriate value?

Reply:
I would be in favor of switching to mcycle. It would make this code not be Ibex-specific (and we can, as such, just delete `ibex.h`, once my link-time PR goes through).

Reply:
Thank you @imphil for looking at the asm closely and @GregAC for pointing this out! I actually verified the behavior using a stopwatch before sending the PR (yes, I wasn't aware that verilator could produce a core trace, still ramping up) and it looked correct. But, I see that the core trace from verilator shows that this snippet takes 4 cycles. So, I added 4 `nop`s (second column is cycle):
```
         177946	     88968	200001a6	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177948	     88969	200001a8	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177950	     88970	200001aa	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177952	     88971	200001ac	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177954	     88972	200001ae	c119	c.beqz	x10,200001b4	 x10:0x0007a120
         177956	     88973	200001b0	1561	c.addi	x10,-8	 x10:0x0007a120 x10=0x0007a118
         177960	     88975	200001b2	bfd5	c.j	200001a6	
         177962	     88976	200001a6	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177964	     88977	200001a8	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177966	     88978	200001aa	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177968	     88979	200001ac	0001	c.addi	x0,0	  x0:0x00000000  x0=0x00000000
         177970	     88980	200001ae	c119	c.beqz	x10,200001b4	 x10:0x0007a118
         177972	     88981	200001b0	1561	c.addi	x10,-8	 x10:0x0007a118 x10=0x0007a110
         177976	     88983	200001b2	bfd5	c.j	200001a6	
```
But, when I run it on the FPGA, it is slower than it should be (~1/3). As far as I can see, RTL and C implementation are in agreement about the clock frequency (50 MHz) so this is probably due to what @GregAC pointed out.

I like the idea of using `mcycle` since it eliminates the need for hand-tuned instructions. I need to look it up.

Reply:
Using the cycle counter resolves this issue.